File: src/LibCut.Api/Infrastructure/ErrorMapping.cs
using System.Text.Json;
using Microsoft.AspNetCore.Mvc;

public static class ErrorMapping
{
    public static IResult ValidationProblem(HttpContext httpContext, LibCutValidationException exception)
    {
        return Results.ValidationProblem(
            errors: exception.ToErrorDictionary().ToDictionary(pair => pair.Key, pair => pair.Value),
            detail: "Correct the fields listed in the errors section and retry the request.",
            title: "Invalid cut optimization request.",
            type: "https://datatracker.ietf.org/doc/html/rfc9110#section-15.5.1",
            statusCode: StatusCodes.Status400BadRequest,
            extensions: CreateExtensions(httpContext));
    }

    public static IResult InvalidJsonProblem(HttpContext httpContext, JsonException exception)
    {
        string path = string.IsNullOrWhiteSpace(exception.Path) ? "$" : exception.Path;
        string location = exception.LineNumber.HasValue && exception.BytePositionInLine.HasValue
            ? $" Line {exception.LineNumber.Value}, byte {exception.BytePositionInLine.Value}."
            : string.Empty;

        var validationException = new LibCutValidationException(
            "Request body contains invalid JSON.",
            new[]
            {
                new LibCutValidationIssue("json", $"Malformed JSON at path '{path}'.{location}".Trim()),
            });

        return ValidationProblem(httpContext, validationException);
    }

    public static IResult UnexpectedFailure(HttpContext httpContext, IHostEnvironment environment, Exception exception)
    {
        var extensions = CreateExtensions(httpContext);
        if (environment.IsDevelopment())
            extensions["exception"] = exception.Message;

        return Results.Problem(
            title: "Cut optimization failed.",
            detail: "Unexpected server error while processing the optimization request.",
            statusCode: StatusCodes.Status500InternalServerError,
            type: "https://datatracker.ietf.org/doc/html/rfc9110#section-15.6.1",
            extensions: extensions);
    }

    private static Dictionary<string, object?> CreateExtensions(HttpContext httpContext)
    {
        return new Dictionary<string, object?>
        {
            ["traceId"] = httpContext.TraceIdentifier,
        };
    }
}

File: libcut_api/src/error_mapping.rs
use axum::http::StatusCode;
use axum::response::{IntoResponse, Response};
use axum::Json;
use libcut_core::error::LibCutValidationError;
use serde::Serialize;
use std::collections::HashMap;

#[derive(Debug, Serialize)]
pub struct ProblemDetail {
    #[serde(rename = "type")]
    pub problem_type: String,
    pub title: String,
    pub status: u16,
    pub detail: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub errors: Option<HashMap<String, Vec<String>>>,
    #[serde(rename = "traceId")]
    pub trace_id: String,
}

impl IntoResponse for ProblemDetail {
    fn into_response(self) -> Response {
        let status =
            StatusCode::from_u16(self.status).unwrap_or(StatusCode::INTERNAL_SERVER_ERROR);
        let mut response = Json(&self).into_response();
        *response.status_mut() = status;
        response.headers_mut().insert(
            axum::http::header::CONTENT_TYPE,
            "application/problem+json"
                .parse()
                .expect("valid header value"),
        );
        response
    }
}

fn new_trace_id() -> String {
    uuid::Uuid::new_v4().to_string()
}

pub fn validation_problem(error: &LibCutValidationError) -> ProblemDetail {
    ProblemDetail {
        problem_type: "https://datatracker.ietf.org/doc/html/rfc9110#section-15.5.1".into(),
        title: "Invalid cut optimization request.".into(),
        status: 400,
        detail: "Correct the fields listed in the errors section and retry the request.".into(),
        errors: Some(error.to_error_dictionary()),
        trace_id: new_trace_id(),
    }
}

pub fn invalid_json_problem(message: &str) -> ProblemDetail {
    let mut errors = HashMap::new();
    errors.insert(
        "json".to_string(),
        vec![format!("Malformed JSON: {}", message)],
    );
    ProblemDetail {
        problem_type: "https://datatracker.ietf.org/doc/html/rfc9110#section-15.5.1".into(),
        title: "Invalid cut optimization request.".into(),
        status: 400,
        detail: "Correct the fields listed in the errors section and retry the request.".into(),
        errors: Some(errors),
        trace_id: new_trace_id(),
    }
}

pub fn empty_body_problem() -> ProblemDetail {
    let mut errors = HashMap::new();
    errors.insert(
        "json".to_string(),
        vec!["Request body is required.".to_string()],
    );
    ProblemDetail {
        problem_type: "https://datatracker.ietf.org/doc/html/rfc9110#section-15.5.1".into(),
        title: "Invalid cut optimization request.".into(),
        status: 400,
        detail: "Correct the fields listed in the errors section and retry the request.".into(),
        errors: Some(errors),
        trace_id: new_trace_id(),
    }
}

pub fn unexpected_failure(_message: &str) -> ProblemDetail {
    ProblemDetail {
        problem_type: "https://datatracker.ietf.org/doc/html/rfc9110#section-15.6.1".into(),
        title: "Cut optimization failed.".into(),
        status: 500,
        detail: "Unexpected server error while processing the optimization request.".into(),
        errors: None,
        trace_id: new_trace_id(),
    }
}
